src/compiler/checker.ts

```diff
@@ -28031,7 +28035,22 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return getNarrowedType(type, instanceType, assumeTrue, /*checkDerived*/ true);
         }
 
-        function getInstanceType(constructorType: Type) {
+        function getInstanceType(constructorType: Type, left: Expression, leftType: Type, right: Expression) {
+            // if rightType is an object type with a custom `[Symbol.hasInstance]` method, and that method has a type
+            // predicate, use the type predicate to perform narrowing. This allows normal `object` types to participate
+            // in `instanceof`, as per Step 2 of https://tc39.es/ecma262/#sec-instanceofoperator.
+            const hasInstanceMethodType = getSymbolHasInstanceMethodOfObjectType(constructorType);
+            if (hasInstanceMethodType) {
+                const syntheticCall = createSyntheticHasInstanceMethodCall(left, leftType, right, constructorType, hasInstanceMethodType);
+                const signature = getEffectsSignature(syntheticCall);
+                const predicate = signature && getTypePredicateOfSignature(signature);
+                if (predicate && predicate.kind === TypePredicateKind.Identifier && predicate.parameterIndex === 0) {
+                    return predicate.type;
+                }
+                if (!isTypeDerivedFrom(constructorType, globalFunctionType)) {

```

Comment:
When does this happen?

Reply:
This is necessary to cover the case originally on original line 28019 where we would have previously exited early: https://github.com/microsoft/TypeScript/pull/55052/files/befa293a880c91110ad57b6d34e924363e8ace85#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8L28019

Instead, we're now performing this check on each constituent of the union, as not all constituents are necessarily going to be function types.